windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -807,14 +823,49 @@ procedure DownloadingState.HandleCheck;
 end;
 
 function DownloadingState.HandleKmShell;
+var
+  FMutex: TKeymanMutex;
 begin
-  // Downloading state, in other process, so continue
+  FMutex := nil;
+  // Whether already downloading in another process or download has failed and
+  // this function has clean up and force a restart, kmshell should continue processing
   Result := kmShellContinue;
+  // Check to ensure a download process is running if not
+  // clean up return to the the idle state and check for updates
+  try
+    FMutex := TKeymanMutex.Create('KeymanDownloading');
+    if FMutex.MutexOwned then
+    begin
+      bucStateContext.RemoveCachedFiles;
+      FreeAndNil(FMutex); // Mutex must be freed before changing state
```

Comment:
I think it would it be cleaner to relinquish ownership without freeing here. Add `TKeymanMutex.Release` function which calls `ReleaseMutex()`